CMakeLists.txt

```diff
@@ -66,28 +63,135 @@ list(APPEND CARTOGRAPHER_INCLUDE_DIRS "${PROTOBUF_INCLUDE_DIR}")
+  list(APPEND ALL_PROTO_HDRS "${CMAKE_BINARY_DIR}/${DIR}/${FIL_WE}.pb.h")
+
+  add_custom_command(
+    OUTPUT "${CMAKE_BINARY_DIR}/${DIR}/${FIL_WE}.pb.cc"
+           "${CMAKE_BINARY_DIR}/${DIR}/${FIL_WE}.pb.h"
+    COMMAND  ${PROTOBUF_PROTOC_EXECUTABLE}
+    ARGS --cpp_out  ${CMAKE_BINARY_DIR} -I
+      ${CMAKE_SOURCE_DIR} ${ABS_FIL}
+    DEPENDS ${ABS_FIL}
+    COMMENT "Running C++ protocol buffer compiler on ${ABS_FIL}"
+    VERBATIM
+  )
+endforeach()
+
+set_source_files_properties(${ALL_PROTO_SRCS} ${ALL_PROTO_HDRS} PROPERTIES GENERATED TRUE)
+list(APPEND ALL_SRCS ${ALL_PROTO_SRCS} ${ALL_PROTO_HDRS})
+
+add_library(cartographer ${ALL_SRCS})
+add_subdirectory("cartographer")
 
-# TODO(hrapp): Replacing this get_property with a generator expression is not
-# very easy: it needs to be expanded into a variable which is then used for
-# CONFIGURE_PACKAGE_CONFIG_FILE - I did not find an easy way to do it. A
-# complicated way is to file(READ the cartographer-config.cmake.in and
-# string(CONFIGURE it, then file(GENERATE the output file. I do not have the
-# stomach to attempt this right now.
-get_property(CARTOGRAPHER_LIBRARY_FILE TARGET cartographer PROPERTY LOCATION)
-get_filename_component(CARTOGRAPHER_LIBRARY_FILE_BASENAME
-  ${CARTOGRAPHER_LIBRARY_FILE} NAME)
-install(
-  FILES
-    ${CARTOGRAPHER_LIBRARY_FILE}
-  DESTINATION
-    lib
+target_include_directories(cartographer SYSTEM PUBLIC
+  "${EIGEN3_INCLUDE_DIR}")
+target_link_libraries(cartographer PUBLIC ${EIGEN3_LIBRARIES})
+
+target_include_directories(cartographer SYSTEM PUBLIC
+  "${CERES_INCLUDE_DIRS}")
+target_link_libraries(cartographer PUBLIC ${CERES_LIBRARIES})
+
+target_include_directories(cartographer SYSTEM PUBLIC
+  "${LUA_INCLUDE_DIR}")
+target_link_libraries(cartographer PUBLIC ${LUA_LIBRARIES})
+
+target_include_directories(cartographer SYSTEM PUBLIC
+  "${Boost_INCLUDE_DIRS}")
+target_link_libraries(cartographer PUBLIC ${Boost_LIBRARIES})
+
+target_link_libraries(cartographer PUBLIC webp)
+
+target_link_libraries(cartographer PUBLIC glog)
+
+target_link_libraries(cartographer PUBLIC gflags)
+
+target_include_directories(cartographer SYSTEM PUBLIC
+  "${CAIRO_INCLUDE_DIRS}")
+target_link_libraries(cartographer PUBLIC ${CAIRO_LIBRARIES})
+
+target_include_directories(cartographer SYSTEM PUBLIC
+  ${PROTOBUF_INCLUDE_DIR})
+# TODO(hrapp): This should not explicitly list pthread and use
+# PROTOBUF_LIBRARIES, but that failed on first try.
+target_link_libraries(cartographer PUBLIC ${PROTOBUF_LIBRARY} pthread)
+
+# Add the binary directory first, so that port.h is included after it has
+# been generated.
+target_include_directories(cartographer PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}>
+    $<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}>
+    $<INSTALL_INTERFACE:include>
 )
 
+target_include_directories(cartographer SYSTEM PRIVATE
+  "${GMOCK_INCLUDE_DIRS}")
+target_link_libraries(cartographer PUBLIC ${GMOCK_LIBRARIES})
```

Comment:
I'm not sure what you mean. You mean pulling out the common CMake code? Because I'd rather not litter all of this code with todos for that.